src/dump.c

```diff
@@ -2381,344 +2386,378 @@ static void jl_insert_methods(jl_array_t *list)
+    htable_new(visited, l);
+    for (i = 0; i < l; i++) {
+        jl_method_instance_t *caller = (jl_method_instance_t*)jl_array_ptr_ref(edges, 2 * i);
+        assert(jl_is_method_instance(caller) && jl_is_method(caller->def.method));
+        jl_array_t *callee_ids = (jl_array_t*)jl_array_ptr_ref(edges, 2 * i + 1);
+        assert(jl_typeis((jl_value_t*)callee_ids, jl_array_int32_type));
+        int valid = 1;
+        if (callee_ids == NULL) {
+            // serializing the edges had failed
+            valid = 0;
+        }
+        else {
+            int32_t *idxs = (int32_t*)jl_array_data(callee_ids);
+            size_t j;
+            for (j = 0; valid && j < idxs[0]; j++) {
+                int32_t idx = idxs[j + 1];
+                valid = jl_array_uint8_ref(valids, idx);
+            }
+            if (!valid && _jl_debug_method_invalidation) {
+                //jl_array_ptr_1d_push(_jl_debug_method_invalidation, (jl_value_t*)m);
+                //loctag = jl_cstr_to_string("jl_method_table_insert"));
+                //jl_array_ptr_1d_push(_jl_debug_method_invalidation, loctag);
+                //loctag = jl_box_int32((int32_t)idxbad);
+                //jl_array_ptr_1d_push(_jl_debug_method_invalidation, loctag);
+                //loctag = jl_box_uint64(jl_worklist_key(serializer_worklist));
+                //jl_array_ptr_1d_push(_jl_debug_method_invalidation, loctag);
+            }
+        }
+        ptrhash_put(visited, caller, (void*)(((char*)HT_NOTFOUND) + valid + 1));
+        //jl_static_show((JL_STREAM*)ios_stderr, (jl_value_t*)caller);
+        //ios_puts(valid ? "valid\n" : "INVALID\n", ios_stderr);
+        // HT_NOTFOUND: valid (no invalid edges)
+        // HT_NOTFOUND + 1: invalid
+        // HT_NOTFOUND + 2: need to scan
+        // HT_NOTFOUND + 3 + depth: in-progress
+    }
+}
+
+// Propagate the result of cycle-resolution to all edges (recursively)
+static void mark_edges_in_worklist(jl_array_t *edges, int idx, htable_t *visited, int found)
+{
+    jl_method_instance_t *caller = (jl_method_instance_t*)jl_array_ptr_ref(edges, idx);
+    int oldfound = (char*)ptrhash_get(visited, caller) - (char*)HT_NOTFOUND;
+    if (oldfound < 3)
+        return; // not in-progress
+    if (!found) {
+        ptrhash_remove(visited, (void*)caller);
+    }
+    else {
+        ptrhash_put(visited, (void*)caller, (void*)((char*)HT_NOTFOUND + 1 + found));
+        if (_jl_debug_method_invalidation) {
+            //jl_array_ptr_1d_push(_jl_debug_method_invalidation, (jl_value_t*)m);
+            //loctag = jl_cstr_to_string("jl_method_table_insert"));
+            //jl_array_ptr_1d_push(_jl_debug_method_invalidation, loctag);
+            //loctag = jl_box_int32((int32_t)idxbad);
+            //jl_array_ptr_1d_push(_jl_debug_method_invalidation, loctag);
+            //loctag = jl_box_uint64(jl_worklist_key(serializer_worklist));
+            //jl_array_ptr_1d_push(_jl_debug_method_invalidation, loctag);
```

Comment:
```suggestion
                jl_array_ptr_1d_push(_jl_debug_method_invalidation, (jl_value_t*)caller);
                loctag = jl_cstr_to_string("jl_verify_methods"));
                jl_array_ptr_1d_push(_jl_debug_method_invalidation, loctag);
                loctag = jl_box_int32((int32_t)idx_caller);
                jl_array_ptr_1d_push(_jl_debug_method_invalidation, loctag);
                loctag = jl_box_uint64(jl_worklist_key(serializer_worklist));
                jl_array_ptr_1d_push(_jl_debug_method_invalidation, loctag);
```
pass idx_caller here